### PR TITLE
[libunwind][libcxx][libcxxabi][compiler-rt-builtins] Fix Exception Handling build for wasm

### PR DESCRIPTION
**Pull request by trcrsired**
The wasm unwind build appears to be dysfunctional, likely because the author has only supplied a customized LLVM build on request, rather than a fully functional patch.

This patch fixes the build

### CODE REVIEW

**Comment by trcrsired** [comment]
i do not know whether old code should be formatted or not, because the LLVM dev said do not format old code

**Comment by trcrsired** [comment]
Fix partial formatting

File: libcxx/include/__exception/exception_ptr.h [hunk 1]
```diff
@@ -36,7 +36,12 @@ struct __cxa_exception;
 _LIBCPP_OVERRIDABLE_FUNC_VIS __cxa_exception* __cxa_init_primary_exception(
     void*,
     std::type_info*,
-    void(
+#  if defined(__USING_WASM_EXCEPTIONS__)
```
Comment (ldionne) [inline libcxx/include/__exception/exception_ptr.h:39]: Let's introduce a typedef name for this function pointer type. Then we can do something like

```c++
#ifdef __USING_WASM_EXCEPTIONS__
using _ExceptionDtor = something;
#else
using _ExceptionDtor = something-else;
#endif
```

and that should be more readable than the current approach.

File: libunwind/include/__libunwind_config.h [hunk 2]
```diff
@@ -180,6 +180,7 @@
 #endif
 #define _LIBUNWIND_HIGHEST_DWARF_REGISTER                                      \
   _LIBUNWIND_HIGHEST_DWARF_REGISTER_LOONGARCH
+#elif defined(__wasm__)
```
Comment (ldionne) [inline libunwind/include/__libunwind_config.h:183]: Why don't we define `_LIBUNWIND_CURSOR_SIZE` and friends on wasm?

File: libunwind/include/libunwind.h [hunk 3]
```diff
@@ -15,6 +15,7 @@
 
 #include <__libunwind_config.h>
 
+#ifndef __wasm__
```
Comment (ldionne) [inline libunwind/include/libunwind.h:18]: What is the purpose of this header if it's entirely empty on wasm?

File: libunwind/src/Unwind-wasm.c [hunk 4]
```diff
@@ -35,7 +32,12 @@ struct _Unwind_LandingPadContext {
 
 // Communication channel between compiler-generated user code and personality
 // function
-thread_local struct _Unwind_LandingPadContext __wasm_lpad_context;
+#if __STDC_VERSION__ >= 202311L
```
Comment (ldionne) [inline libunwind/src/Unwind-wasm.c:35]: We always compile the built library with C++23, so I don't understand this change.

File: libunwind/src/Unwind-wasm.c [hunk 5]
```diff
@@ -10,14 +10,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if __STDC_VERSION__ < 202311L
```
Comment (ldionne) [inline libunwind/src/Unwind-wasm.c:13]: This change seems unrelated to the rest of the patch.

File: libunwind/src/libunwind.cpp [hunk 6]
```diff
@@ -12,6 +12,7 @@
 #include <libunwind.h>
 
 #include "config.h"
+#ifndef __wasm__
```
Comment (ldionne) [inline libunwind/src/libunwind.cpp:15]: Similar question as above here. If this file is basically empty on wasm, we should instead avoid adding it in the CMakeLists.txt.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libunwind/include/libunwind.h (hunk 3 above)
Comment (trcrsired) [reply libunwind/include/libunwind.h:18]: it does not compile. all the definitions are not defined

File: libunwind/src/Unwind-wasm.c (hunk 4 above)
Comment (trcrsired) [reply libunwind/src/Unwind-wasm.c:35]: thread.h does not exist.

File: libunwind/src/Unwind-wasm.c (hunk 5 above)
Comment (trcrsired) [reply libunwind/src/Unwind-wasm.c:13]: stdbool.h removed in C23

File: libunwind/src/libunwind.cpp (hunk 6 above)
Comment (trcrsired) [reply libunwind/src/libunwind.cpp:15]: the problem is that the debugging code needs in libunwind

File: libunwind/include/__libunwind_config.h (hunk 2 above)
Comment (trcrsired) [reply libunwind/include/__libunwind_config.h:183]: What is the correct value here? I mean if you give users a number and they may make wrong assumptions here.

File: libcxx/include/__exception/exception_ptr.h (hunk 1 above)
Comment (trcrsired) [reply libcxx/include/__exception/exception_ptr.h:39]: yeah. but here's a problem. __USING_WASM_EXCEPTIONS__ is not defined by clang 19 even with -fwasm-exceptions enabled. i suggest to just use __wasi__ macro, instead of __USING_WASM_EXCEPTIONS__

File: libunwind/src/Unwind-wasm.c (hunk 4 above)
Comment (trcrsired) [reply libunwind/src/Unwind-wasm.c:35]: > ldionne

it is a c file. not a .cpp file. it compiles with C compiler not C++

File: libunwind/src/libunwind.cpp (hunk 6 above)
Comment (trcrsired) [reply libunwind/src/libunwind.cpp:15]: i have included apis with minimal code.
https://github.com/llvm/llvm-project/blob/b5fee4962c347016a732a1310275651a8c74f23a/libunwind/src/libunwind.cpp#L442
It needs logAPIs in the file

File: libcxx/include/__exception/exception_ptr.h (hunk 1 above)
Comment (trcrsired) [reply libcxx/include/__exception/exception_ptr.h:39]: ok i will try to fix it.

Comment (trcrsired) [reply libcxx/include/__exception/exception_ptr.h]: Fixed

**Comment by trcrsired** [comment]
@ldionne Hi. I think i have fixed as many issues as I can. can you review it again? Thank you so much!

File: libunwind/src/libunwind_ext.h [hunk 7]
```diff
@@ -12,6 +12,7 @@
 #ifndef __LIBUNWIND_EXT__
 #define __LIBUNWIND_EXT__
 
+#ifndef __wasm__
```
Comment (EricWF) [inline libunwind/src/libunwind_ext.h:15]: This feels like the wrong way to do this. We don't typically just if-def out a header for a particular platform.

Maybe track down the users of this header, and make them work under WASM?

File: libunwind/src/libunwind.cpp (hunk 6 above)
Comment (EricWF) [inline libunwind/src/libunwind.cpp:15]: It seems like we simply don't want this file to build under WASM.

**Review by EricWF — CHANGES_REQUESTED** [review]
I think there are some pretty non-idiomatic workarounds for WASM here. I think in general this approach could use a more structured approach to handling where WASM differs.

Also, I know we have the mandatory code-formatter now, but a lot of this diff is just formatting noise. Could we find a way to extract that out?

File: libunwind/src/libunwind.cpp (hunk 6 above)
Comment (trcrsired) [reply libunwind/src/libunwind.cpp:15]: then it has to change cmake file, which is annoying for users.

File: libunwind/src/libunwind_ext.h (hunk 7 above)
Comment (trcrsired) [reply libunwind/src/libunwind_ext.h:15]: The problem is that __WASM_EXCEPTIONS__ macro is not defined by clang, so using any other the header without __wasm__ will trigger ODR violation. Remote unwinding is not a thing for wasm, so i do not see what's wrong here

File: libunwind/src/libunwind.cpp (hunk 6 above)
Comment (EricWF) [reply libunwind/src/libunwind.cpp:15]: I feel like we could detect `__wasm__` or similar in CMake. And it feels like the right tool, though maybe not the most trivial to implement.

Comment (trcrsired) [reply libunwind/src/libunwind.cpp:15]: > I feel like we could detect `__wasm__` or similar in CMake. And it feels like the right tool, though maybe not the most trivial to implement.

Tbh i would always avoid wrapping around cmake, if platform macro works out.

File: libunwind/src/assembly.h [hunk 8]
```diff
@@ -15,6 +15,7 @@
 #ifndef UNWIND_ASSEMBLY_H
 #define UNWIND_ASSEMBLY_H
 
+#ifndef __wasm__
```
Comment (EricWF) [inline libunwind/src/assembly.h:18]: Again, this is not OK.

First, this file just defines macros, so even under WASM it should be safe to include.
If you're hitting the `#error` case in the `#else`, then add an `#elif __wasm__`.

While his style of code may not cause issues for you, because you don't use it. It makes things more complicated for every change that actually uses these files.

File: libunwind/src/libunwind_ext.h (hunk 7 above)
Comment (EricWF) [reply libunwind/src/libunwind_ext.h:15]: I would much rather you guard the usage of this header in each file.

That way it's clear to the reader in context that the include is unused under a certain configuration.
If the entire file contains no WASM code, then it probably shouldn't mention WASM at all.

Can you please explain what the `__WASM_EXCEPTIONS__` or `__USING_WASM_EXCEPTIONS__` was?

File: libunwind/src/cet_unwind.h [hunk 9]
```diff
@@ -39,3 +40,5 @@ extern void *__libunwind_cet_get_registers(unw_cursor_t *);
 extern void *__libunwind_cet_get_jump_target(void);
```
Comment (EricWF) [inline libunwind/src/cet_unwind.h:40]: What is the problem with exposing these two declarations, even if they're not defined?

**Comment by EricWF** [comment]
@trcrsired Could you please describe what bits of libunwind WASM uses? It seems to me almost all of the functionality is stripped out?

Could you also provide instructions to build and test libunwind under WASM so we could add CI coverage? And so I can test this change myself.

File: libunwind/src/cet_unwind.h (hunk 9 above)
Comment (trcrsired) [reply libunwind/src/cet_unwind.h:40]: there are no registers. so what would you expect for cet_get_jump_target? there is no setjmp/longjmp

**Comment by trcrsired** [comment]
> @trcrsired Could you please describe what bits of libunwind WASM uses? It seems to me almost all of the functionality is stripped out?
> 
> Could you also provide instructions to build and test libunwind under WASM so we could add CI coverage? And so I can test this change myself.

ok. i will add CI

File: libunwind/src/libunwind_ext.h (hunk 7 above)
Comment (trcrsired) [reply libunwind/src/libunwind_ext.h:15]: wasm has its own exception handling mechanism. i do not know why __USING_WASM_EXCEPTION__ is not defined, because the author of the code did not change the upstream code correctly i guess.

**Comment by trcrsired** [comment]
@EricWF You need to build it with wasi-libc

```cpp
mkdir -p "$CURRENTTRIPLEPATH/build/runtimes"
cd $CURRENTTRIPLEPATH/build/runtimes

cmake $LLVMPROJECTPATH/runtimes \
	-GNinja -DCMAKE_BUILD_TYPE=Release \
	-DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_ASM_COMPILER=clang \
	-DCMAKE_SYSROOT=$SYSROOTPATH -DCMAKE_INSTALL_PREFIX=$RUNTIMESINSTALLPATH \
	-DCMAKE_C_COMPILER_TARGET=$TARGETTRIPLE -DCMAKE_CXX_COMPILER_TARGET=$TARGETTRIPLE -DCMAKE_ASM_COMPILER_TARGET=$TARGETTRIPLE \
	-DCMAKE_C_COMPILER_WORKS=On -DCMAKE_CXX_COMPILER_WORKS=On -DCMAKE_ASM_COMPILER_WORKS=On \
	-DCMAKE_SYSTEM_PROCESSOR=$TARGETTRIPLE_CPU \
	-DCMAKE_SYSTEM_NAME=Generic \
	-DLLVM_ENABLE_RUNTIMES=$EHBUILDLIBS \
	-DLIBCXXABI_SILENT_TERMINATE=On \
	-DLIBCXX_CXX_ABI=libcxxabi \
	-DLIBCXX_ENABLE_SHARED=Off \
	-DLIBCXX_ABI_VERSION=1 \
	-DLIBCXX_CXX_ABI_INCLUDE_PATHS="${LLVMPROJECTPATH}/libcxxabi/include" \
	${THREADS_FLAGS} \
	-DLIBCXX_ENABLE_EXCEPTIONS=${ENABLE_EH} \
	-DLIBCXXABI_ENABLE_EXCEPTIONS=${ENABLE_EH} \
	-DLIBCXX_ENABLE_RTTI=${ENABLE_EH} \
	-DLIBCXXABI_ENABLE_RTTI=${ENABLE_EH} \
	-DLLVM_ENABLE_ASSERTIONS=Off -DLLVM_INCLUDE_EXAMPLES=Off -DLLVM_ENABLE_BACKTRACES=Off -DLLVM_INCLUDE_TESTS=Off -DLIBCXX_INCLUDE_BENCHMARKS=Off \
	-DLIBCXX_ENABLE_SHARED=Off -DLIBCXXABI_ENABLE_SHARED=Off \
	-DLIBUNWIND_ENABLE_SHARED=Off ${EH_FLAGS}
ninja
ninja install/strip
```

File: compiler-rt/lib/builtins/fp_compare_impl.inc [hunk 10]
```diff
@@ -12,7 +12,7 @@
 // functions. We need to ensure that the return value is sign-extended in the
 // same way as GCC expects (since otherwise GCC-generated __builtin_isinf
 // returns true for finite 128-bit floating-point numbers).
-#ifdef __aarch64__
+#if defined(__aarch64__) || defined(__wasm__)
```
[the remark this answers is not in the document]
Comment (arichardson) [reply compiler-rt/lib/builtins/fp_compare_impl.inc:15]: Agreed, please leave an explicit comment here for WASM. When I ran into this issue it was really difficult to track down, so I really appreciate having explicit comments that show the rational for the special cases.

File: libcxx/include/__exception/exception_ptr.h [hunk 11]
```diff
@@ -29,22 +29,21 @@
 
 namespace __cxxabiv1 {
 
+#    if defined(__wasm__)
+typedef void* (*__libcpp_exception_destructor_func)(void*);
```
Comment (arichardson) [inline libcxx/include/__exception/exception_ptr.h:33]: Could keep the existing comment here `// In Wasm, a destructor returns its argument` to explain why it is different?

File: libcxxabi/include/cxxabi.h [hunk 12]
```diff
@@ -98,73 +89,58 @@ extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_NORETURN void __cxa_pure_virtual(void);
...
-#if defined(_LIBCXXABI_GUARD_ABI_ARM)
-extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_ALWAYS_COLD int __cxa_guard_acquire(uint32_t *);
-extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_ALWAYS_COLD void __cxa_guard_release(uint32_t *);
-extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_ALWAYS_COLD void __cxa_guard_abort(uint32_t *);
-#else
-extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_ALWAYS_COLD int __cxa_guard_acquire(uint64_t *);
-extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_ALWAYS_COLD void __cxa_guard_release(uint64_t *);
-extern _LIBCXXABI_FUNC_VIS _LIBCXXABI_ALWAYS_COLD void __cxa_guard_abort(uint64_t *);
-#endif
+#  if defined(_LIBCXXABI_GUARD_ABI_ARM)
```
Comment (arichardson) [inline libcxxabi/include/cxxabi.h:92]: There are a lot of formatting changes here that make this hard to review - could you commit the clang-format changes first and then rebase this PR on top?

**Review by arichardson — COMMENTED** [review]
I started reviewing this but all the unrelated formatting changes made this quite difficult. Would appreciate a minimal diff with just the changes that are needed for WASM.

**Review by ldionne — CHANGES_REQUESTED** [review]
I think support for wasm mandates a short RFC, like most projects to support new platforms that require non-trivial changes. There's not a huge amount of changes in this patch, but it's also unclear to me that this is sufficient to do anything.

In particular, the following should be covered:
- What are the particularities of this platform, if any
- Who is supporting this platform and what is the expected support story
- What is the story for testing. How do you run tests on that platform? Do we need to make changes to the libc++ conformance test suite for things to work?

That would be the minimum. We have a policy of not accepting changes to add support for new platforms for platforms that are not officially supported. It would be amazing to support wasm and I'm sure it can be done, but we need to do it the right way, not only whip out a patch and call it a day.

I know we've accepted some isolated wasm changes in the past so this may sound surprising, however given the scope of the changes here I think this deserves to be done properly.

File: libunwind/src/Unwind-wasm.c [hunk 13]
```diff
@@ -10,15 +10,13 @@
 //
 //===----------------------------------------------------------------------===//
 
+// clang-format off
```
Comment (arichardson) [inline libunwind/src/Unwind-wasm.c:13]: Why is this needed? Adding another newline should avoid the reordering. Alternatively, config.h should include stdbool.h to be self-contained.

**Review by arichardson — CHANGES_REQUESTED** [review]
There are tons of unrelated changes that make this really hard to review. Please do the formatting part of the changes separately and address the outstanding comments.

**Comment by arichardson** [comment]
Could you address the raised issues instead of constantly updating the branch? This wastes CI resources and spams everyone subscribed to this PR. Please feel free to reopen it once you have addressed the issues.
